File: 2150-kth-smallest-product-of-two-sorted-arrays/2150-kth-smallest-product-of-two-sorted-arrays.py
class Solution:
    def kthSmallestProduct(self, nums1: List[int], nums2: List[int], k: int) -> int:
        # Separate negatives and non-negatives, flip negatives to positives
        def split_and_reverse_negatives(arr):
            neg = []
            pos = []
            for x in arr:
                if x < 0:
                    neg.append(-x)  # Flip negative to positive
                else:
                    pos.append(x)
            neg.reverse()  # reverse to keep ascending order
            return neg, pos

        # Count number of products ≤ m using two pointers
        def countProductsLEQ(arr1, arr2, max_product):
            count = 0
            j = len(arr2) - 1
            for x in arr1:
                while j >= 0 and x * arr2[j] > max_product:
                    j -= 1
                count += j + 1
            return count

        # Separate and flip negative numbers for easier handling
        neg1, pos1 = split_and_reverse_negatives(nums1)
        neg2, pos2 = split_and_reverse_negatives(nums2)

        # Total number of negative products: one element is negative, other is positive
        total_negatives = len(neg1) * len(pos2) + len(pos1) * len(neg2)

        resultSign = 1
        if k > total_negatives:
            # kth product is positive
            k -= total_negatives
        else:
            # kth product is negative => we reverse sign later
            k = total_negatives - k + 1
            resultSign = -1
            neg2, pos2 = pos2, neg2  # switch arrays to find the neg product as pos

        # Binary search to find the kth smallest *positive* product
        left = 0
        right = 10**10

        while left < right:
            mid = (left + right) // 2
            # Count how many products are ≤ mid
            count = countProductsLEQ(neg1, neg2, mid) + countProductsLEQ(pos1, pos2, mid)
            if count >= k:
                right = mid
            else:
                left = mid + 1

        return resultSign * left

File: 2150-kth-smallest-product-of-two-sorted-arrays/2150-kth-smallest-product-of-two-sorted-arrays.java
class Solution {
    public long kthSmallestProduct(int[] nums1, int[] nums2, long k) {
        // Step 1: Separate nums1 and nums2 into positive and reversed-negative arrays
        List<Long> neg1 = new ArrayList<>();
        List<Long> pos1 = new ArrayList<>();
        List<Long> neg2 = new ArrayList<>();
        List<Long> pos2 = new ArrayList<>();

        for (int x : nums1) {
            if (x < 0) neg1.add((long)(-x)); // flip sign
            else pos1.add((long)x);
        }
        Collections.reverse(neg1); // to maintain sorted order

        for (int x : nums2) {
            if (x < 0) neg2.add((long)(-x)); // flip sign
            else pos2.add((long)x);
        }
        Collections.reverse(neg2); // to maintain sorted order

        long totalNegativePairs = 1L * neg1.size() * pos2.size() + 1L * pos1.size() * neg2.size();
        int sign = 1;

        if (k > totalNegativePairs) {
            // kth product is positive
            k -= totalNegativePairs;
        } else {
            // kth product is negative
            k = totalNegativePairs - k + 1;
            sign = -1;
            // swap neg2 and pos2 for correct pairing
            List<Long> temp = neg2;
            neg2 = pos2;
            pos2 = temp;
        }

        long left = 0, right = (long)1e10;

        while (left < right) {
            long mid = (left + right) / 2;
            long count = countLEQ(neg1, neg2, mid) + countLEQ(pos1, pos2, mid);

            if (count >= k) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return sign * left;
    }

    // Count how many products a * b <= maxProduct using two-pointer technique
    private long countLEQ(List<Long> A, List<Long> B, long maxProduct) {
        long count = 0;
        int j = B.size() - 1;

        for (long a : A) {
            while (j >= 0 && a * B.get(j) > maxProduct) {
                j--;
            }
            count += (j + 1);
        }

        return count;
    }
}
